source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -157,8 +158,24 @@ Network::IoResult SslSocket::doRead(Buffer::Instance& read_buffer) {
     if (slices_to_commit > 0) {
       read_buffer.commit(slices, slices_to_commit);
       if (callbacks_->shouldDrainReadBuffer()) {
-        callbacks_->setReadBufferReady();
-        keep_reading = false;
+        // Verify that SSL_get_read_ahead is disabled. SSL_pending does not provide an accurate
+        // answer when read-ahead is enabled. As far as we can tell BoringSSL does not implement
+        // read-ahead, so the odds of us ever failing this sanity check are effectively zero.
+        ASSERT(!SSL_get_read_ahead(rawSsl()));
+
+        // Query the SSL implementation for the number of bytes available for immediate read from
+        // internal buffers, and do one last read iteration if bytes are available. This is
+        // important to ensure read resumption works correctly after calls to
+        // Network::Connection::readDisable().
+        int pending_bytes = SSL_pending(rawSsl());
+        ASSERT(pending_bytes < 16 * 1024, "SSL record should be at most 16KB");
```

Comment:
Should this be <= ?

Reply:
The record size is at most 16KB, but at least 1 byte from the record must have been consumed by an earlier SSL_read, so SSL_pending can return at most 16KB-1

Reply:
OK makes sense. Can you add a small comment and adjust the assert error to s/at most/less than?